src/Agent.Worker/ContainerOperationProvider.cs

```diff
@@ -271,137 +327,127 @@ public async Task StartContainerAsync(IExecutionContext executionContext, object
+                }
 
-            // Add the new created user to the new created sudo group.
-            int execUsermodExitCode = await _dockerManger.DockerExec(executionContext, container.ContainerId, string.Empty, $"usermod -a -G azure_pipelines_sudo {containerUserName}");
-            if (execUsermodExitCode != 0)
-            {
-                throw new InvalidOperationException($"Docker exec fail with exit code {execUsermodExitCode}");
-            }
+                // Add the new created user to the new created sudo group.
+                int execUsermodExitCode = await _dockerManger.DockerExec(executionContext, container.ContainerId, string.Empty, $"usermod -a -G azure_pipelines_sudo {containerUserName}");
+                if (execUsermodExitCode != 0)
+                {
+                    throw new InvalidOperationException($"Docker exec fail with exit code {execUsermodExitCode}");
+                }
 
-            // Allow the new sudo group run any sudo command without providing password.
-            int execEchoExitCode = await _dockerManger.DockerExec(executionContext, container.ContainerId, string.Empty, $"su -c \"echo '%azure_pipelines_sudo ALL=(ALL:ALL) NOPASSWD:ALL' >> /etc/sudoers\"");
-            if (execUsermodExitCode != 0)
-            {
-                throw new InvalidOperationException($"Docker exec fail with exit code {execEchoExitCode}");
+                // Allow the new sudo group run any sudo command without providing password.
+                int execEchoExitCode = await _dockerManger.DockerExec(executionContext, container.ContainerId, string.Empty, $"su -c \"echo '%azure_pipelines_sudo ALL=(ALL:ALL) NOPASSWD:ALL' >> /etc/sudoers\"");
+                if (execUsermodExitCode != 0)
+                {
+                    throw new InvalidOperationException($"Docker exec fail with exit code {execEchoExitCode}");
+                }
             }
 #endif
         }
 
-        public async Task StopContainerAsync(IExecutionContext executionContext, object data)
+        private async Task StopContainerAsync(IExecutionContext executionContext, ContainerInfo container)
         {
             Trace.Entering();
             ArgUtil.NotNull(executionContext, nameof(executionContext));
-            ContainerInfo container = data as ContainerInfo;
             ArgUtil.NotNull(container, nameof(container));
 
-            if (!string.IsNullOrEmpty(container.ContainerId))
-            {
-                executionContext.Output($"Stop and remove container: {container.ContainerDisplayName}");
-
-                int rmExitCode = await _dockerManger.DockerRemove(executionContext, container.ContainerId);
-                if (rmExitCode != 0)
-                {
-                    executionContext.Error($"Docker rm fail with exit code {rmExitCode}");
-                }
+            executionContext.Output($"Stop and remove container: {container.ContainerDisplayName}");
```

Comment:
why no longer check ContainerId? if containerid is null then, we never start the container.

Reply:
Fixed